resotolib/resotolib/graph/__init__.py

```diff
@@ -166,16 +199,81 @@ def add_edge(self, src, dst, **attr):
             except Exception:
                 log.exception(
                     (
-                        f"Unhandeled exception while telling {dst.rtdname}"
+                        f"Unhandled exception while telling {dst.rtdname}"
                         f" that {src.rtdname} was added as a predecessor"
                     )
                 )
+        return return_key
+
+    def remove_node(self, node: BaseResource):
+        super().remove_node(node)
+
+    def remove_edge(
+        self,
+        src: BaseResource,
+        dst: BaseResource,
+        key: EdgeKey = None,
+        edge_type: EdgeType = None,
+    ):
+        if edge_type is None:
+            edge_type = EdgeType.default
+        if key is None:
+            key = EdgeKey(src=src, dst=dst, edge_type=edge_type)
+        super().remove_edge(src, dst, key=key)
+
+    def predecessors(self, node: BaseResource, edge_type: EdgeType = None):
+        if edge_type is None:
+            edge_type = EdgeType.default
+        for predecessor in super().predecessors(node):
+            key = (predecessor, node, edge_type)
+            if self.has_edge(predecessor, node, key=key):
+                yield predecessor
+
+    def successors(self, node: BaseResource, edge_type: EdgeType = None):
+        if edge_type is None:
+            edge_type = EdgeType.default
+        for successor in super().successors(node):
+            key = (node, successor, edge_type)
+            if self.has_edge(node, successor, key=key):
+                yield successor
+
+    def ancestors(self, node: BaseResource, edge_type: EdgeType = None):
+        return networkx.algorithms.dag.ancestors(
+            self.edge_type_subgraph(edge_type), node
+        )
 
-    def remove_node(self, *args, **kwargs):
-        super().remove_node(*args, **kwargs)
+    def descendants(self, node: BaseResource, edge_type: EdgeType = None):
+        return networkx.algorithms.dag.descendants(
+            self.edge_type_subgraph(edge_type), node
+        )
 
-    def remove_edge(self, *args, **kwargs):
-        super().remove_edge(*args, **kwargs)
+    def edge_type_subgraph(self, edge_type: EdgeType = None):
+        if edge_type is None:
+            edge_type = EdgeType.default
+        edges = []
+        for edge in self.edges(keys=True):
+            if len(edge) == 3:
+                key: EdgeKey = edge[2]
+                if key.edge_type == edge_type:
+                    edges.append(edge)
+        return self.edge_subgraph(edges)
+
+    def is_dag(self) -> bool:
```

Comment:
Only naming: readers might think that if this returns true the whole multigraph is a dag, which is not true.

Reply:
good point, renamed